File: tool-router/tool_router/services/hierarchical_router.py
from __future__ import annotations

import time

from tool_router.config import Settings
from tool_router.models import Environment, ParsedIntent, ResourceRecord, ResourceType
from tool_router.services.bm25 import Bm25Index
from tool_router.services.registry import RegistryStore, clean_domain, infer_domain_groups, DOMAIN_GROUPS


class HierarchicalRouteResult:
    def __init__(
        self,
        domain_groups: list[str],
        domains: list[str],
        capabilities: list[str],
        resources: list[ResourceRecord],
    ) -> None:
        self.domain_groups = domain_groups
        self.domains = domains
        self.capabilities = capabilities
        self.resources = resources


class HierarchicalRouter:
    """四级分层路由器：DomainGroup → Domain → Capability → Resource。

    检索空间逐层缩小，禁止全量遍历：
      Level-1 DomainGroup（BM25 降级兜底）
      Level-2 Domain（BM25 降级兜底）
      Level-3 Capability
      Level-4 Resource（前置过滤 + 跨域图谱扩展）
    """

    def __init__(self, registry: RegistryStore, settings: Settings) -> None:
        self.registry = registry
        self.settings = settings
        self.route_cache: dict[tuple[str, str, str, str, str], tuple[float, list[str]]] = {}
        # DomainGroup 级 BM25（Level-1 路由降级用）
        self._group_bm25 = Bm25Index()
        for group, domains in DOMAIN_GROUPS.items():
            self._group_bm25.add(group, " ".join(domains))
        # Domain 级 BM25（Level-2 路由降级用）
        self._domain_bm25 = Bm25Index()
        for group, domains in DOMAIN_GROUPS.items():
            for domain in domains:
                self._domain_bm25.add(domain, domain)

    def route(self, tenant_id: str, environment: Environment, intent: ParsedIntent) -> HierarchicalRouteResult:
        # ===== Level-1: DomainGroup =====
        groups = self._route_to_group(intent)
        # ===== Level-2: Domain =====
        domains = self._route_to_domain(intent, groups)
        # ===== Neo4j 跨领域依赖跳转 =====
        domains = list(dict.fromkeys(domains + self._graph_expand(domains)))
        # ===== Level-3: Capability =====
        capabilities = self._resolve_capabilities(intent, domains)

        cache_key = (
            tenant_id,
            environment.value,
            ",".join(groups),
            ",".join(domains),
            ",".join(capabilities),
        )
        cached = self.route_cache.get(cache_key)
        if cached is not None and cached[0] > time.time():
            cached_resources = [
                record
                for resource_id in cached[1]
                if (record := self.registry.get(resource_id)) is not None
            ]
            if cached_resources:
                return HierarchicalRouteResult(groups, domains, capabilities, cached_resources)

        # ===== Level-4: Resource（按 domain 分片 + 前置过滤 + 图谱补全） =====
        resources = self.registry.route_search(
            tenant_id=tenant_id,
            environment=environment.value,
            domain_groups=groups,
            domains=domains,
            capabilities=capabilities,
            read_only=intent.query_constraints.read_only,
            auth_level=intent.query_constraints.auth_level,
            file_type=intent.query_constraints.file_type,
        )
        # 精确 capability 无结果时，同 domain 降级（保证不返回空）
        if not resources and capabilities:
            resources = self.registry.route_search(
                tenant_id=tenant_id,
                environment=environment.value,
                domain_groups=groups,
                domains=domains,
                capabilities=[],
                read_only=intent.query_constraints.read_only,
                auth_level=intent.query_constraints.auth_level,
                file_type=intent.query_constraints.file_type,
            )
        self.route_cache[cache_key] = (
            time.time() + self.settings.route_cache_ttl_seconds,
            [resource.level1.resource_id for resource in resources],
        )
        return HierarchicalRouteResult(groups, domains, capabilities, resources)

    # ===== Level-1: DomainGroup 路由（BM25 降级） =====
    def _route_to_group(self, intent: ParsedIntent) -> list[str]:
        """从 domain_candidates 上推到 DomainGroup；无命中时用 BM25 兜底。"""
        groups: list[str] = []
        for domain in intent.domain_candidates:
            for group in infer_domain_groups(domain, ResourceType.tool):
                if group not in groups:
                    groups.append(group)
        if groups:
            return groups
        # 降级：BM25 在 DomainGroup 索引上匹配
        hits = self._group_bm25.search(intent.intent, top_k=2)
        return [h[0] for h in hits] if hits else ["general"]

    # ===== Level-2: Domain 路由（BM25 降级） =====
    def _route_to_domain(self, intent: ParsedIntent, groups: list[str]) -> list[str]:
        """从 DomainGroup 下推到具体 Domain；优先 intent 命中的 domain。"""
        domains: list[str] = []
        for group in groups:
            for domain in DOMAIN_GROUPS.get(group, []):
                if domain not in domains:
                    domains.append(domain)
        if intent.domain_candidates:
            intersection = [d for d in intent.domain_candidates if d in domains]
            if intersection:
                return intersection
            return intent.domain_candidates
        if domains:
            return domains
        # 降级：BM25 在 Domain 索引上匹配
        hits = self._domain_bm25.search(intent.intent, top_k=3)
        return [h[0] for h in hits] if hits else ["misc"]

    # ===== Level-3: Capability 解析 =====
    @staticmethod
    def _resolve_capabilities(intent: ParsedIntent, domains: list[str]) -> list[str]:
        suffix = intent.primary_capability.split(".", 1)[1] if "." in intent.primary_capability else "general"
        out = [intent.primary_capability.lower()]
        for domain in domains:
            candidate = f"{domain}.{suffix}".lower()
            if candidate not in out:
                out.append(candidate)
            general = f"{domain}.general"
            if general not in out:
                out.append(general)
        return out

    # ===== Neo4j 跨领域依赖跳转 =====
    def _graph_expand(self, domains: list[str]) -> list[str]:
        """通过 combine_with 关系发现关联领域，补齐跨域候选。"""
        expanded: list[str] = []
        for record in self.registry.list_records():
            if clean_domain(record.level1.domain) not in domains:
                continue
            for edge in self.registry.list_graph_edges(record.level1.resource_id, {"combine_with"}):
                target = self.registry.get(edge.target_id)
                if target is None:
                    continue
                cleaned = clean_domain(target.level1.domain)
                if cleaned not in domains and cleaned not in expanded:
                    expanded.append(cleaned)
        return expanded

File: server/src/tools/tool-search/hierarchical-router/hierarchical-router.ts
import { createClient, type RedisClientType } from "redis";

import type { ParsedIntent, QueryConstraints } from "../intent-router/intent-router.js";
import type { ResourceRecord } from "../registry/models.js";
import type { ToolRegistryStore } from "../registry/store.js";
import { getCurrentToolRegistryEnvironment } from "../registry/registry-service.js";

export type DomainGroupName =
  | "communication"
  | "productivity"
  | "system"
  | "knowledge"
  | "commerce"
  | "world"
  | "misc";

export type HierarchicalRouteInput = {
  tenant_id: string;
  parsed_intent: ParsedIntent;
  max_resources?: number;
};

export type HierarchicalRouteResult = {
  domain_groups: DomainGroupName[];
  domains: string[];
  capabilities: string[];
  resources: ResourceRecord[];
  cache_hit: boolean;
  filtered_count: number;
};

export type HierarchicalRouterOptions = {
  redisUrl?: string;
  cacheTtlMs?: number;
  expandDomains?: (domains: string[], intent: ParsedIntent) => Promise<string[]>;
};

const DEFAULT_CACHE_TTL_MS = 20_000;
const CACHE_PREFIX = "tool:route:";

const DOMAIN_GROUPS: Record<DomainGroupName, string[]> = {
  communication: ["phone", "agent", "calendar", "reminder"],
  productivity: ["calendar", "reminder", "file", "notes", "self"],
  system: ["desktop", "browser", "embodiment", "clock", "aip"],
  knowledge: ["search", "browser", "weather", "file", "notes"],
  commerce: ["shopping", "wallet", "budget"],
  world: ["world", "agent"],
  misc: ["misc"],
};

/**
 * Phase-2 Hierarchical Router.
 *
 * 只通过 registry route index 读取 domain/capability 切片，禁止调用全资源列表做检索。
 */
export class HierarchicalRouter {
  private readonly redisUrl: string | undefined;
  private readonly cacheTtlMs: number;
  private readonly expandDomains?: (domains: string[], intent: ParsedIntent) => Promise<string[]>;
  private redis: RedisClientType | null = null;
  private redisConnectPromise: Promise<void> | null = null;

  constructor(
    private readonly store: ToolRegistryStore,
    options?: HierarchicalRouterOptions,
  ) {
    this.redisUrl =
      (options?.redisUrl ?? process.env.AGENT_REDIS_URL?.trim()) || undefined;
    this.cacheTtlMs = Math.max(1000, options?.cacheTtlMs ?? DEFAULT_CACHE_TTL_MS);
    this.expandDomains = options?.expandDomains;
  }

  async route(input: HierarchicalRouteInput): Promise<HierarchicalRouteResult> {
    const maxResources = Math.max(1, Math.min(500, input.max_resources ?? 100));
    const domains = await this.resolveDomains(input.parsed_intent);
    const domainGroups = resolveDomainGroups(domains);
    const capabilities = resolveCapabilities(input.parsed_intent, domains);
    const cacheKey = buildCacheKey(input.tenant_id, domains, capabilities);

    const cached = await this.readHotCache(cacheKey);
    if (cached.length > 0) {
      const records = await this.recordsFromIds(cached, input.parsed_intent.query_constraints);
      if (records.length > 0) {
        return {
          domain_groups: domainGroups,
          domains,
          capabilities,
          resources: records.slice(0, maxResources),
          cache_hit: true,
          filtered_count: records.length,
        };
      }
    }

    const env = getCurrentToolRegistryEnvironment();
    let resources = await this.store.listRecordsByRoute({
      tenantId: input.tenant_id,
      env,
      domains,
      capabilities,
      limit: maxResources,
    });

    // 精确 capability 无结果时，只在同 domain 内降级，仍然不进入全局扫描。
    if (resources.length === 0 && capabilities.length > 0) {
      resources = await this.store.listRecordsByRoute({
        tenantId: input.tenant_id,
        env,
        domains,
        limit: maxResources,
      });
    }

    const filtered = filterByConstraints(resources, input.parsed_intent.query_constraints);
    await this.writeHotCache(cacheKey, filtered.map((r) => r.level1.resource_id));

    return {
      domain_groups: domainGroups,
      domains,
      capabilities,
      resources: filtered,
      cache_hit: false,
      filtered_count: resources.length - filtered.length,
    };
  }

  private async resolveDomains(intent: ParsedIntent): Promise<string[]> {
    const base = unique(
      intent.domain_candidates.length > 0 ? intent.domain_candidates : ["misc"],
    );
    if (!this.expandDomains) return base;
    try {
      return unique([...(await this.expandDomains(base, intent)), ...base]);
    } catch (e) {
      console.warn("[tool-search:route] graph domain expansion failed", e);
      return base;
    }
  }

  private async recordsFromIds(
    ids: string[],
    constraints: QueryConstraints,
  ): Promise<ResourceRecord[]> {
    const records: ResourceRecord[] = [];
    for (const id of ids) {
      const record = await this.store.getRecord(id);
      if (record) records.push(record);
    }
    return filterByConstraints(records, constraints);
  }

  private async readHotCache(key: string): Promise<string[]> {
    const redis = await this.getRedis();
    if (!redis) return [];
    try {
      const raw = await redis.get(key);
      if (!raw) return [];
      const ids = JSON.parse(raw) as unknown;
      return Array.isArray(ids) ? ids.map(String) : [];
    } catch (e) {
      console.warn("[tool-search:route] hot cache read failed", e);
      return [];
    }
  }

  private async writeHotCache(key: string, ids: string[]): Promise<void> {
    if (ids.length === 0) return;
    const redis = await this.getRedis();
    if (!redis) return;
    try {
      await redis.set(key, JSON.stringify(ids), { PX: this.cacheTtlMs });
    } catch (e) {
      console.warn("[tool-search:route] hot cache write failed", e);
    }
  }

  private async getRedis(): Promise<RedisClientType | null> {
    if (!this.redisUrl) return null;
    if (this.redis?.isOpen) return this.redis;
    if (!this.redisConnectPromise) {
      const client = createClient({ url: this.redisUrl });
      client.on("error", (err) =>
        console.warn("[tool-search:route] redis error", err),
      );
      this.redis = client as RedisClientType;
      this.redisConnectPromise = client.connect().then(
        () => undefined,
        (e) => {
          console.warn("[tool-search:route] redis connect failed", e);
          this.redis = null;
        },
      );
    }
    await this.redisConnectPromise;
    return this.redis?.isOpen ? this.redis : null;
  }
}

function resolveDomainGroups(domains: string[]): DomainGroupName[] {
  const out: DomainGroupName[] = [];
  for (const [group, groupDomains] of Object.entries(DOMAIN_GROUPS) as Array<
    [DomainGroupName, string[]]
  >) {
    if (domains.some((d) => groupDomains.includes(d))) out.push(group);
  }
  return out.length ? out : ["misc"];
}

function resolveCapabilities(intent: ParsedIntent, domains: string[]): string[] {
  const out = new Set<string>();
  if (intent.primary_capability) out.add(intent.primary_capability);
  for (const domain of domains) {
    out.add(`${domain}.general`);
    const prefix = intent.primary_capability.split(".")[1];
    if (prefix) out.add(`${domain}.${prefix}`);
  }
  return Array.from(out);
}

function filterByConstraints(
  records: ResourceRecord[],
  constraints: QueryConstraints,
): ResourceRecord[] {
  const requestedAuth = constraints.auth_level;
  return records.filter((record) => {
    if (record.level1.status !== "online") return false;
    if (!authAllows(record.auth_level, requestedAuth)) return false;
    if (constraints.read_only && isLikelyWriteResource(record)) return false;
    if (
      constraints.file_type &&
      record.level1.tags.length > 0 &&
      !record.level1.tags.some((tag) => tag.toLowerCase() === constraints.file_type)
    ) {
      return false;
    }
    return true;
  });
}

function authAllows(resourceAuth: string, requestedAuth: string): boolean {
  if (resourceAuth === "guest") return true;
  if (resourceAuth === "default") return requestedAuth !== "guest";
  return requestedAuth === "admin";
}

function isLikelyWriteResource(record: ResourceRecord): boolean {
  const text = [
    record.level1.name,
    record.level1.description,
    ...record.level1.tags,
    ...record.level2.use_cases,
  ]
    .join(" ")
    .toLowerCase();
  return /delete|remove|write|send|create|update|transfer|pay|删除|发送|创建|更新|转账|支付/.test(
    text,
  );
}

function buildCacheKey(
  tenantId: string,
  domains: string[],
  capabilities: string[],
): string {
  return `${CACHE_PREFIX}${tenantId}:${domains.join(",")}:${capabilities.join(",")}`;
}

function unique(values: string[]): string[] {
  return Array.from(new Set(values.map((v) => v.trim()).filter(Boolean)));
}
